### PR TITLE
Made some changes to Array, and shuffled some string/array functions …

### PR DESCRIPTION
**Pull request by TurkeyMan**
…around.

### CODE REVIEW

File: src/urt/array.d [hunk 1]
```diff
@@ -128,7 +179,16 @@ struct Array(T, size_t EmbedCount = 0)
 
     // constructors
 
-    this(this) @disable;
+    // TODO: DELETE POSTBLIT!
+    this(this)
```
Comment (megaeels) [inline src/urt/array.d:183]: What is this doing?

File: src/urt/string/string.d [hunk 2]
```diff
@@ -61,7 +79,12 @@ nothrow @nogc:
 
     this(typeof(null)) pure
     {
-        ptr = null;
+        this.ptr = null;
+    }
+
+    this(ref inout typeof(this) rhs) inout pure
```
Comment (megaeels) [inline src/urt/string/string.d:85]:
```suggestion
    this(ref in typeof(this) rhs) inout pure
```

File: src/urt/string/string.d [hunk 3]
```diff
@@ -192,7 +215,14 @@ nothrow @nogc:
     BaseString!char _super;
     alias _super this;
 
-    this(this) @disable;
+    // TODO: DELETE POSTBLIT!
```
Comment (megaeels) [inline src/urt/string/string.d:218]: Why is this needed?

File: src/urt/string/string.d [hunk 4]
```diff
@@ -13,9 +13,27 @@ enum MaxStringLen = 0x3FFF;
...
+        else
+            buffer[0] = cast(char)(lit.length & 0x7F);
+        buffer[LenBytes .. LenBytes + lit.length] = lit[];
+        static if (zeroTerminate)
+            buffer[$-1] = '\0'; // add a zero terminator for good measure
+        return buffer;
+    }();
+    private __gshared immutable literal = LiteralData;
+
+    enum String StringLit = String(literal.ptr + LenBytes, null);
```
Comment (megaeels) [inline src/urt/string/string.d:36]: Does this work if the data isn't null terminated?

File: src/urt/array.d [hunk 5]
```diff
@@ -22,6 +22,57 @@ T[] pop(T)(ref T[] arr, ptrdiff_t n)
...
+}
+
+ref inout(T) popBack(T)(ref inout(T)[] buffer) pure nothrow @nogc
+{
+	debug assert(buffer.length > 0);
+	buffer = buffer.ptr[0..buffer.length - 1];
+	return buffer.ptr[buffer.length];
+}
+
+inout(T)[] takeFront(T)(ref inout(T)[] s, size_t count) pure nothrow @nogc
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/urt/array.d:44]: It used to be a string function, then I generalised it...

File: src/urt/array.d [hunk 6]
```diff
@@ -22,6 +22,57 @@ T[] pop(T)(ref T[] arr, ptrdiff_t n)
...
+	assert(count <= s.length);
+	inout(T)[] t = s.ptr[0 .. count];
+	s = s.ptr[count .. s.length];
+	return t;
+}
+
+ref inout(T)[N] takeFront(size_t N, T)(ref inout(T)[] s) pure nothrow @nogc
+{
+	assert(N <= s.length);
+	inout(T)* t = s.ptr;
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/urt/array.d:55]: It seemed pointless to perform a redundant bounds-check, even though it will optimise away.

File: src/urt/array.d (hunk 1 above)
Comment (TurkeyMan) [reply src/urt/array.d:183]: I need to look into why this freaked out when I deleted the postblit (replaced by copy constructors years ago). I wondered if I bumped into some kind of bug...
I just implemented a postblit to move on, but it needs to investigation.

File: src/urt/string/string.d (hunk 4 above)
Comment (TurkeyMan) [reply src/urt/string/string.d:36]: It doesn't care about null termination. I just added the option in case the string needs to be passed to a C function without building a null-terminated temp.

File: src/urt/string/string.d (hunk 3 above)
Comment (TurkeyMan) [reply src/urt/string/string.d:218]: Same reason as above...
I didn't take the time to investigate.

File: src/urt/array.d [hunk 7]
```diff
@@ -22,6 +22,57 @@ T[] pop(T)(ref T[] arr, ptrdiff_t n)
 //Slice<T> take(ptrdiff_t n)
 //Slice<T> drop(ptrdiff_t n)
 
+bool empty(T)(T[] arr) pure nothrow @nogc
+{
+    return arr.length == 0;
+}
+
+ref inout(T) popFront(T)(ref inout(T)[] buffer) pure nothrow @nogc
```
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/urt/array.d:30]: Yeah... I've been stressing on this all over the shop.
I'm not sure that a tiny computer can suffer the code bloat of doing a bounds check in EVERY array access. I definitely want to catch errors in dev, but I'm not sure that the release build needs it... but that depends on some policies about how arrays are used.
I'd like to find patterns that hoist bounds checks upwards, and not embed them in all the micro-operations everywhere.

Not sure if that can work, but it needs a big conversation.

File: src/urt/array.d (hunk 5 above)
[the remark this answers is not in the document]
Comment (TurkeyMan) [reply src/urt/array.d:44]: `buffer` is not a good name anymore either because they're `T`s... so, `items`? `array`? :/

File: src/urt/string/string.d (hunk 2 above)
Comment (TurkeyMan) [reply src/urt/string/string.d:85]: I can't assign a mutable pointer if `this` is `const`.

https://dlang.org/spec/const3.html#inout

`inout` is the stupidest named keyword in D, but it essentially means "transfer const-ness"; how that got named is anybodies guess!
